[DA] Add BatchDelinearization to DependenceAnalysis

sebpop opened the pull request:
This adds BatchDelinearization support to DependenceAnalysis with a new
constructor that takes a target loop:

      DependenceInfo(Function *F, AAResults *AA, ScalarEvolution *SE,
                     LoopInfo *LI, Loop &L);

When created with a target loop, DependenceInfo performs batch
delinearization upfront for all memory accesses in the loop, ensuring
consistent array dimensions across all accesses to the same base pointer.

kasuga-fj reviewed (commented):
- DA and Delinearization are already tightly coupled. I think it's better to separate them properly first. Maybe adding a feature like `BatchDelinearization` to Delinearization (like `BatchAliasAnalysis`) would be a good idea.
- Supporting both fixed-size and parametric batch delinearization at the same time makes the change complex. I think it's better to first support only fixed-size batch delinearization.
    - Notably, due to the recent changes in the validation logic of delinearization, currently parametric delinearization will fail almost all cases. This is because we cannot prove that the product of estimated dimensions (e.g., `N * M` for `A[N][M]`) doesn't overflow. See ##169048 and #169902 for details.

sebpop commented on polly/include/polly/ScopDetection.h:
@alexey-bataev asked to "Use SmallDenseMap, it improves compile time", and thus I also impacted this change to Polly from where I got the caching mechanism.

sebpop replied:
I removed those changes.

sebpop committed:
[DA] Add BatchDelinearization to DependenceAnalysis

This adds BatchDelinearization support to DependenceAnalysis with a new
constructor that takes a target loop:

  DependenceInfo(Function *F, AAResults *AA, ScalarEvolution *SE,
                 LoopInfo *LI, Loop &L);

When created with a target loop, DependenceInfo performs batch
delinearization upfront for all memory accesses in the loop, ensuring
consistent array dimensions across all accesses to the same base pointer.

kasuga-fj commented on llvm/include/llvm/Analysis/Delinearization.h:
```diff
@@ -164,6 +171,84 @@ bool getIndexExpressionsFromGEP(ScalarEvolution &SE,
+  SmallDenseMap<const SCEVUnknown *, const SCEV *, 8> ElementSizes;
+
+  /// Flag indicating whether the cache has been populated.
+  bool Populated = false;
```
Consider using [RAII](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization) rather than holding such a flag.

Meinersbur reviewed (commented):
Have you considered makeing `BatchDelineraization` an pass manager analysis? That would already do the caching.

kasuga-fj reviewed (commented):
If I remember correctly, the original motivation is to cache the delinearization results within a single transformation (e.g., loop-interchange), not share them across multiple passes. So making `BatchDelinearization` an analysis pass would be irrelevant to the original intent. Also, as the name "batch" suggests, I had assumed that delinearization would be performed up front for all memory accesses in the target loop.

Current approach makes the result query-order dependent, which I think is not ideal. I had the following algorithm in mind:

```cpp
BatchDelinearization::BatchDelinearization(...) {
  Map<SCEV *, Vector<Instruction *>> BasePtr2Insts;
  for (Instruction &I : instructions in the target loop) {
    // Collect all memory accesses and group them by their base pointers, then store.
  }

  Map<SCEV *, Vector<SCEV *>> BasePtr2EstimatedSize;
  for (auto &[Base, Insts] : BasePtr2Insts) {
    Vector<Vector<SCEV *>> EstimatedSizes;
    for (Instruction *I : Insts) {
      // Estimate the array sizes for each memory access and store.
    }

    // Somehow "generalize" the estimated sizes. 
    Vector<SCEV *> Sizes = GeneralizeEstimatedSizes(EstimatedSizes);
    Cache[Base] = Sizes;
  }
}
```

kasuga-fj commented on llvm/include/llvm/Analysis/DDG.h:
```diff
@@ -287,13 +288,12 @@ template <typename NodeType> class DependenceGraphInfo {
-  // queried it is recomputed using @DI.
-  const DependenceInfo DI;
+  // DependenceInfo for computing memory dependencies on demand.
+  // Owned by this graph to ensure proper lifetime.
```
Avoid irrelevant changes.

kasuga-fj commented on llvm/include/llvm/Analysis/DDG.h:
```diff
@@ -287,13 +288,12 @@ template <typename NodeType> class DependenceGraphInfo {
-  // A special node in the graph that has an edge to every connected component of
-  // the graph, to ensure all nodes are reachable in a graph walk.
+  // A special node in the graph that has an edge to every connected component
+  // of the graph, to ensure all nodes are reachable in a graph walk.
```
Avoid irrelevant changes.

kasuga-fj commented on llvm/test/Transforms/LICM/lnicm.ll:
```diff
@@ -3,9 +3,6 @@
 
-; XFAIL: *
-; Loop interchange currently fails due to a failure in dependence analysis.
-
```
Keep the test marked as XFAIL as it is.

kasuga-fj reviewed (commented):
The motivation behind batch delinearization is to improve its accuracy, right? If so, the test case `llvm/test/Analysis/DependenceAnalysis/Invariant.ll` is a typical example, where `rr[j][j]` is interpreted as an access to 1D array. It is now marked as XFAIL, but batch delinearization should return it to PASS, without changing any CHECK lines.

Also note that `BatchDelinearization` doesn't run in the tests under `llvm/test/Analysis/DependenceAnalysis/`. Probably `DependenceAnalysisPrinterPass` and/or `dumpExampleDependence` need to be updated, or another way is stop passing a `Loop` object to initialize `BatchDelinearization`.
